query/file_log_watcher.go

```diff
@@ -0,0 +1,164 @@
+	return f.fsWatcher.Events
+}
+
+func (f *FileLogWatcher) Error() chan error {
+	return f.fsWatcher.Errors
+}
+
+func (f *FileLogWatcher) Add(name string) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	err := f.fsWatcher.Add(name)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *FileLogWatcher) Remove(name string) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	err := f.fsWatcher.Remove(name)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func NewFileLogWatcher(e *Executor, path string, logger *zap.Logger, format string) *FileLogWatcher {
+	// FileWatcher is not meant to be used with interactive shell sessions so
+	// console as a logging format will not be supported.
+	if format == "console" {
+		logger.Error("unknown logging format", zap.String("format", format), zap.String("path", path))
+		return nil
+	}
+
+	if format == "" || format == "auto" {
+		format = "logfmt"
+	}
+
+	logFile, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0644)
+	if err != nil {
+		logger.Error("failed to open log file", zap.Error(err))
+		return nil
+	}
+
+	existingCore := logger.Core()
+	encoder, err := l.NewEncoder(format)
+	if err != nil {
+		logger.Error("failed to create log encoder", zap.Error(err), zap.String("format", format), zap.String("path", path))
+		return nil
+	}
+
+	fileCore := zapcore.NewCore(
+		encoder,
+		zapcore.Lock(logFile),
+		zapcore.InfoLevel,
+	)
+
+	newCore := zapcore.NewTee(existingCore, fileCore)
+	logger = zap.New(newCore)
+
+	fsWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logger.Error("failed to create fsnotify watcher", zap.Error(err), zap.String("path", path))
+		return nil
+	}
+
+	return &FileLogWatcher{
+		logger:          logger,
+		path:            path,
+		currFile:        logFile,
+		executor:        e,
+		formatterConfig: format,
+		mu:              sync.Mutex{},
+		fsWatcher:       fsWatcher,
+	}
+}
+
+func (f *FileLogWatcher) GetLogger() *zap.Logger {
+	return f.logger
+}
+
+func (f *FileLogWatcher) GetLogPath() string {
+	return f.path
+}
+
+func (f *FileLogWatcher) FileChangeCapture() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if err := f.currFile.Sync(); err != nil {
+		return err
+	}
+	if err := f.currFile.Close(); err != nil {
+		return err
+	}
+
+	logFile, err := os.OpenFile(f.path, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0644)
+	if err != nil {
+		f.logger.Error("failed to open log file", zap.Error(err), zap.String("path", f.path))
+		return nil
+	}
+	f.currFile = logFile
+	existingCore := f.logger.Core()
+	encoder, err := l.NewEncoder(f.formatterConfig)
+	if err != nil {
+		return err
+	}
+
+	fileCore := zapcore.NewCore(
+		encoder,
+		zapcore.Lock(logFile),
+		zapcore.InfoLevel,
+	)
+
+	newCore := zapcore.NewTee(existingCore, fileCore)
+	f.logger = zap.New(newCore)
+
+	if err := f.fsWatcher.Add(f.path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (f *FileLogWatcher) Close() error {
+	if err := f.currFile.Sync(); err != nil {
```

Comment:
- Do we need mutex protection here?
- We should always close the file, even if the sync fails.
- We should close the `fsWatcher` even if the sync or close fails
- `errors.Join` is your friend.